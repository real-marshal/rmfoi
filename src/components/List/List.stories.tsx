import type { Meta, Story } from '@storybook/react'
import { List, ListProps } from './List'
import { Basic as BasicListItem, Selected } from './ListItem/ListItem.stories'

export default {
  title: 'Common/List/List',
  component: List,
} as Meta

const Template: Story<ListProps> = (args) => <List {...args} />

export const Basic = Template.bind({})
Basic.args = {
  children: [
    <Selected {...Selected.args} key={1}>
      Selected List Item 1
    </Selected>,
    <BasicListItem {...BasicListItem.args} key={2}>
      Basic List Item 2
    </BasicListItem>,
    <BasicListItem {...BasicListItem.args} key={3}>
      Basic List Item 3
    </BasicListItem>,
  ],
}
