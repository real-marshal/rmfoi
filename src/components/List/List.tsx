import { dt } from '@/features/Theming'
import { css } from '@emotion/react'
import styled from '@emotion/styled'
import { forwardRef, ReactNode } from 'react'
import { ListItem } from './ListItem'

interface ListProps {
  children: ReactNode
  align?: 'left' | 'right'
}

const StyledUL = styled.ul<Pick<ListProps, 'align'>>`
  background: linear-gradient(${dt.theme.background}, ${dt.theme.background}) padding-box,
    ${dt.theme.gradientBlue};
  border: 4px solid transparent;
  border-radius: 10px;
  z-index: ${dt.z.menu};
  padding: 0;
  width: fit-content;
  position: absolute;

  ${({ align }) =>
    align === 'right' &&
    css`
      right: 0;
    `}
`

const List = forwardRef<HTMLUListElement, ListProps>(({ children, align = 'left' }, ref) => (
  <StyledUL ref={ref} align={align}>
    {children}
  </StyledUL>
))

List.displayName = 'List'

const ListCompound = Object.assign(List, { ListItem })

export { ListCompound as List }
export type { ListProps }
