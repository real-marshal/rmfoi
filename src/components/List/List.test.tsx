import { composeStories } from '@storybook/react'
import { render } from '@testing-library/react'
import * as stories from './List.stories'

describe('List component', () => {
  const testStories = Object.values(composeStories(stories)).map(
    (Story) => [Story.storyName, Story] as const
  )

  it.each(testStories)('renders %s story', (_, Story) => {
    const view = render(<Story />)
    expect(view.baseElement).toMatchSnapshot()
  })
})
