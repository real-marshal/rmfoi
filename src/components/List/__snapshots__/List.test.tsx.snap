// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component renders Basic story 1`] = `
.emotion-0 {
  background: linear-gradient(var(--dt-theme-background), var(--dt-theme-background)) padding-box,var(--dt-colors-gradient);
  border: 4px solid transparent;
  border-radius: 10px;
  padding: 0;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  position: absolute;
}

.emotion-2 {
  color: var(--dt-theme-primary);
  background-color: var(--dt-theme-background);
  list-style: none;
  padding: var(--dt-padding-xl) var(--dt-padding-xl2);
  cursor: pointer;
  color: var(--dt-theme-background);
  background-color: var(--dt-theme-primary);
}

.emotion-2:first-of-type {
  border-radius: 6px 6px 0 0;
}

.emotion-2:last-of-type {
  border-radius: 0 0 6px 6px;
}

.emotion-2:hover {
  color: var(--dt-theme-background);
  background-color: var(--dt-theme-primary);
}

.emotion-4 {
  color: var(--dt-theme-primary);
  background-color: var(--dt-theme-background);
  list-style: none;
  padding: var(--dt-padding-xl) var(--dt-padding-xl2);
  cursor: pointer;
}

.emotion-4:first-of-type {
  border-radius: 6px 6px 0 0;
}

.emotion-4:last-of-type {
  border-radius: 0 0 6px 6px;
}

.emotion-4:hover {
  color: var(--dt-theme-background);
  background-color: var(--dt-theme-primary);
}

<body>
  <div>
    <ul
      class="emotion-0 emotion-1"
    >
      <li
        class="emotion-2 emotion-3"
      >
        Selected List Item 1
      </li>
      <li
        class="emotion-4 emotion-3"
      >
        Basic List Item 2
      </li>
      <li
        class="emotion-4 emotion-3"
      >
        Basic List Item 3
      </li>
    </ul>
  </div>
</body>
`;
